### PR TITLE
Implement StandardGameState, CellsUnderThreat and PieceMoves.

### PR DESCRIPTION
**Pull request by SherlockKA**
- [x] StandardGameState
- [x] CellsUnderThreat
- [x] PieceMoves
- [x] run tests automatically
Tests are written, but I could not run them automatically.
- [x] update csproj file  

TODO:
- [ ] define the style guide
- [ ] complete StandardGame class

### CODE REVIEW

File: Program.cs [hunk 1]
```diff
@@ -1,3 +1,16 @@
...
+
+    class Program
+    {
+        static void Main(string[] args)
+        {
+        }
+
+
+    }
+}
```
Comment (amirov-m) [inline Program.cs:16]: All source code files should end with a new line.
https://www.quora.com/Why-is-it-a-common-practice-to-put-a-new-line-at-the-end-of-a-source-code-file


You may configure your IDE to add a finish new line automatically if missing.

Please fix in other files too.

File: engine/CellsUnderThreat.cs [hunk 2]
```diff
@@ -0,0 +1,402 @@
...
+            };
+            return FindCells(cell,
+                             shifts,
+                             pieceCells,
+                             enemyPieceCells,
+                             OnBoard,
+                             oneShift:false);
+        }
+
+        /// <summary>
```
Comment (amirov-m) [inline engine/CellsUnderThreat.cs:272]: Для чего именно такие большие комментарии с хэдерами? Довольно сложно читается.

File: engine/IFENSerializable.cs [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+    {
+        /*
+            The interface represents the serialization/deserialization to/from FEN notation.
+
+            Please refer to the FEN notation (https://en.wikipedia.org/wiki/Forsyth%E2%80%93Edwards_Notation). 
+        */
+
+        // Serialize to FEN notation.
+        string SerializeToFEN(T objectToSerialize);
+        // Deserialize from FEN notation.
```
Comment (amirov-m) [inline engine/IFENSerializable.cs:16]: Add an empty line between method declarations.

File: engine/IFENSerializable.cs [hunk 4]
```diff
@@ -0,0 +1,19 @@
+using System;
+
+namespace LightChess
+{
+
```
Comment (amirov-m) [inline engine/IFENSerializable.cs:5]: Redundant new line.

File: engine/IStandardGameLogic.cs [hunk 5]
```diff
@@ -0,0 +1,27 @@
+using System;
+using System.Collections.Generic;
+
+namespace LightChess
+{
+
```
Comment (amirov-m) [inline engine/IStandardGameLogic.cs:6]: Redundant new line.

File: engine/PieceMoves.cs [hunk 6]
```diff
@@ -0,0 +1,290 @@
...
+                && !pieceCells.Contains(cellOneMoveForward)
+                && !enemyPieceCells.Contains(celltwoMovesForward))
+            {
+                cellsNext.Add(celltwoMovesForward);
+            }
+
+            return cellsNext;
+        }
+    }
+
```
Comment (amirov-m) [inline engine/PieceMoves.cs:289]: Redundant new line.

File: entities/StandardBoard.cs [hunk 7]
```diff
@@ -0,0 +1,176 @@
...
+
+        public StandardBoard()
+        {
+            PositionToPiece = new Dictionary<Cell, Piece>();
+        }
+
+        public StandardBoard(Dictionary<Cell, Piece> positionToPiece)
+        {
+            // Deep copy of dictionary.
+            PositionToPiece = positionToPiece.ToDictionary(entry => entry.Key,
```
Comment (amirov-m) [inline entities/StandardBoard.cs:29]: What did you mean by "Deep copy" here exactly?

I can see that you make a shallow copy as objects used for keys and values reused. And that's fine given that the object are immutable as I understand.

Another way to achieve the same is:
`PositionToPiece = new Dictionary<Cell, Piece>(positionToPiece);`

File: entities/StandardBoard.cs [hunk 8]
```diff
@@ -0,0 +1,176 @@
...
+        /// <summary>
+        /// Set piece at cell.
+        /// </summary>
+        /// <param name="cell">
+        /// The cell.
+        /// </param>
+        /// <param name="piece">
+        /// The piece.
+        /// </param>
+        public void SetPiece(Cell cell,
```
Comment (amirov-m) [inline entities/StandardBoard.cs:116]: Why are parameters placed on different lines? I believe there is enough space for both parameters.

File: entities/StandardGameState.cs [hunk 9]
```diff
@@ -0,0 +1,313 @@
...
+
+        public StandardGameState(
+            StandardBoard board,
+            Color color,
+            List<Castle> avaialbleCastles,
+            Cell? enPassantCell,
+            int halfmoveNumber,
+            int fullmoveNumber
+        )
+        {
```
Comment (amirov-m) [inline entities/StandardGameState.cs:35]: I prefer having an opening curly bracket on a previous line, but it is up to us to decide which style guide we follow.

File: entities/StandardGameState.cs [hunk 10]
```diff
@@ -0,0 +1,313 @@
...
+                nextBoard = PerformCastle(gameState.Board, move);
+            }
+            // Enpassant move.
+            else if (gameState.EnPassantCell != null 
+                    &&  move.EndCell == (Cell)gameState.EnPassantCell)
+            {
+                nextBoard = PerformEnPassantMove(gameState.Board, move);
+            }
+            // Simple move.
+            else
```
Comment (amirov-m) [inline entities/StandardGameState.cs:77]: Here also I would prefer writing `} else {`, but again, we can agree on a style guide.

File: test/entities/StandardGameState_Test.cs [hunk 11]
```diff
@@ -0,0 +1,106 @@
...
+                  "h1", "h8")]
+        // Capture of queen rook to check an update of castles.                  
+        [TestCase("r3k2r/8/8/8/8/8/8/R3K2R w KkQq - 2 20", 
+                  "R3k2r/8/8/8/8/8/8/4K2R b Kk - 0 21",
+                  "a1", "a8")]
+
+        // 2. Black moves
+        // King castle.       
+        [TestCase("r3k2r/p2p1p1p/n2B4/1p1NP1pP/Pp4P1/3P1Q2/2P5/R3K2R b KkQq a3 2 20", 
+                  									 "r4rk1/p2p1p1p/n2B4/1p1NP1pP/Pp4P1/3P1Q2/2P5/R3K2R w KQ - 3 21",
```
Comment (amirov-m) [inline test/entities/StandardGameState_Test.cs:56]: Why are parameters so spaced?

**Review by amirov-m — CHANGES_REQUESTED** [review]
Comments about style guide.

File: Program.cs (hunk 1 above)
Comment (SherlockKA) [reply Program.cs]: Fixed!

File: engine/IFENSerializable.cs (hunk 3 above)
Comment (SherlockKA) [reply engine/IFENSerializable.cs:16]: Fixed!

File: engine/IStandardGameLogic.cs (hunk 5 above)
Comment (SherlockKA) [reply engine/IStandardGameLogic.cs]: Fixed!

File: entities/StandardBoard.cs (hunk 7 above)
Comment (SherlockKA) [reply entities/StandardBoard.cs]: Fixed!

File: Program.cs (hunk 1 above)
Comment (SherlockKA) [reply Program.cs]: Fixed!

File: entities/GameState.cs [hunk 12]
```diff
@@ -0,0 +1,30 @@
...
+            FEN notation (https://en.wikipedia.org/wiki/Forsyth%E2%80%93Edwards_Notation). 
+        */
+
+        // The board contains the piece locations/cells.
+        StandardBoard board;
+        // The color defines the current turn.
+        Color ActiveColor;
+        // The list contains available castles.
+        List<Castle> AvaialbleCastles;
+        // The list contains avaialble enpassant moves.
```
Comment (cardinal-evgeniy) [inline entities/GameState.cs:22]: typo in "avaialble"

File: entities/GameState.cs [hunk 13]
```diff
@@ -0,0 +1,30 @@
...
+            P.S. The game state contains fields/properties inherited from
+            FEN notation (https://en.wikipedia.org/wiki/Forsyth%E2%80%93Edwards_Notation). 
+        */
+
+        // The board contains the piece locations/cells.
+        StandardBoard board;
+        // The color defines the current turn.
+        Color ActiveColor;
+        // The list contains available castles.
+        List<Castle> AvaialbleCastles;
```
Comment (cardinal-evgeniy) [inline entities/GameState.cs:21]: AvailableCastleMoves & AvailableEnPassantMoves

File: entities/PieceType.cs [hunk 14]
```diff
@@ -0,0 +1,19 @@
+using System;
+
+namespace LightChess
+{
+
+    public enum PieceType
```
Comment (cardinal-evgeniy) [inline entities/PieceType.cs:5]: change order to 

Pawn,
Knight,
Bishop,
Rook,
Queen,
Knight

File: entities/IBoard.cs [hunk 15]
```diff
@@ -0,0 +1,28 @@
...
+{
+
+    public interface IBoard
+    {
+        /*
+            The interface represents the required board functionality.  
+        */
+
+        // The method checks if cell coordinates are valid.
+        bool OnBoard(Cell cell);
```
Comment (cardinal-evgeniy) [inline entities/IBoard.cs:13]: IsOnBoard

File: entities/StandardGameState.cs [hunk 16]
```diff
@@ -0,0 +1,313 @@
...
+        /// <param name="board">
+        /// The start/initial board.
+        /// </param>
+        /// <param name="move">
+        /// The move to perform.
+        /// </param>        
+        /// <returns>
+        /// The next board with performed move. 
+        /// </returns>
+        private static StandardBoard PerformMove(StandardBoard board,
```
Comment (cardinal-evgeniy) [inline entities/StandardGameState.cs:205]: maybe rename to MakeMove

File: engine/CellsUnderThreat.cs (hunk 2 above)
Comment (SherlockKA) [reply engine/CellsUnderThreat.cs]: Fixed!

**Review by amirov-m — CHANGES_REQUESTED** [review]

File: entities/Color.cs [hunk 17]
```diff
@@ -0,0 +1,25 @@
...
+        */
+        
+        White,
+        Black
+    }
+
+    static class ColorExtension
+    {
+
+        public static Color Change(this Color color)
```
[the remark this answers is not in the document]
Comment (SherlockKA) [reply entities/Color.cs]: How?

File: GameLogic/engine/StandardGame.cs [hunk 18]
```diff
@@ -0,0 +1,186 @@
...
+            );
+            return gameState;
+        }
+
+
+        /// <summary>
+        /// TODO: not finished.
+        /// </summary>
+        public bool IsMate(StandardGameState gameState)
+        {
```
Comment (amirov-m) [inline GameLogic/engine/StandardGame.cs:74]: Тут ведь ещё должна быть проверка по идее, что в данный момент тоже есть шах? Или просто пока не закончено?

File: entities/StandardGameState.cs (hunk 16 above)
Comment (SherlockKA) [reply entities/StandardGameState.cs]: I prefer PerformMove. @amirov-m please choose either PerformMove or MakeMove.

File: entities/GameState.cs (hunk 13 above)
Comment (SherlockKA) [reply entities/GameState.cs]: Fixed!

File: GameLogic/engine/IStandardGameLogic.cs [hunk 19]
```diff
@@ -0,0 +1,28 @@
...
+namespace GameLogic
+{
+    // The interface represents the logic of chess game. 
+    public interface IStandardGameLogic
+    {
+        // Checks if the mate occurs at current game state.
+        bool IsMate(StandardGameState gameState);
+
+        // Checks if the check occurs at current game state.
+        bool IsCheck(StandardGameState gameState);
```
[the remark this answers is not in the document]
Comment (SherlockKA) [reply GameLogic/engine/IStandardGameLogic.cs:13]: Yes

File: GameLogic/engine/IStandardGameLogic.cs [hunk 20]
```diff
@@ -0,0 +1,28 @@
...
+                                   Move move);
+
+        // Checks if the game state is possible.
+        bool IsGameStateValid(StandardGameState gameState);
+
+        // Find all moves. Moves can be not valid.
+        List<Move> FindAllMoves(StandardGameState gameState);
+
+        // Find all cells under threat produced by enemy
+        List<Cell> FindAllCellsUndeThreat(StandardGameState gameState);
```
[the remark this answers is not in the document]
Comment (SherlockKA) [reply GameLogic/engine/IStandardGameLogic.cs:26]: Fixed!

File: GameLogic/engine/IStandardGameLogic.cs [hunk 21]
```diff
@@ -0,0 +1,28 @@
...
+    // The interface represents the logic of chess game. 
+    public interface IStandardGameLogic
+    {
+        // Checks if the mate occurs at current game state.
+        bool IsMate(StandardGameState gameState);
+
+        // Checks if the check occurs at current game state.
+        bool IsCheck(StandardGameState gameState);
+
+        // Makes a move if possible, otherwise return null.
```
[the remark this answers is not in the document]
Comment (SherlockKA) [reply GameLogic/engine/IStandardGameLogic.cs:15]: Fixed!

File: GameLogic.Tests/engine/CellsUnderThreat_Test.cs [hunk 22]
```diff
@@ -0,0 +1,84 @@
...
+using System.Linq;
+using System.Collections.Generic;
+using NUnit.Framework;
+
+namespace GameLogic.Tests
+{
+    [TestFixture]
+    public class CellsUnderThreat_Test
+    {
+
```
[the remark this answers is not in the document]
Comment (SherlockKA) [reply GameLogic.Tests/engine/CellsUnderThreat_Test.cs:11]: Fixed!

**Review by amirov-m — APPROVED** [review]

**Pull request merged**